### ISSUE TITLE
Call methods on configuration of objects

### ISSUE THREAD

**Issue by vtvz** [yiisoft/yii2#15085]
I can configure objects using `Yii::configure` method where `key` is a `param`.

I suggest to add syntax to call methods on configuration. Example:

```php
[
    'a' => 'foo',
    'b' => 'bar',
    '!c' => 'car',
    '!d' => ['m', 'r'],
]
```

This configuration sets `$obj->a = 'foo'` and `$obj->b = 'bar'` AND calls `$obj->c('car')` and `$obj->d('m', 'r')`

It can be special useful in DI container if a want to write configuration of not framework classes:

```php
[
    'class' => SomeExternalClassOfNotYii2Framework::class,
    '!setLogger' => Instance::of('logger'),
]
``` 

Ready to make a PR for this functionality

**Comment by SamMousa** [comment]
Why not provide a closure as definition instead?

**Comment by vtvz** [comment]
@SamMousa example please. I don't fully understand what do you mean.

**Comment by SamMousa** [comment]
You can use a Closure (inline function) as definition for your DI configuration.

**Comment by samdark** [comment]
Yes. To me closures looks a bit nicer than properties defined with a special syntax...

**Comment by vtvz** [comment]
Closures is not a config. But configs you can merge as arrays (and it's really handy). But closures (no)?

**Comment by samdark** [comment]
Yes, you can't merge closures.

**Comment by vtvz** [comment]
And you can `invoke` methods through DI conteiner to resolve it's own dependencies. I think it's could very powerful

**Comment by samdark** [comment]
I still don't like the syntax :(

**Comment by vtvz** [comment]
Syntax can be another. I offered the idea. And waiting for your opinions and offers. And if maintainers will decide to include this feature to framework I'll be really glad to help with code and tests.

If not... Let it be

Another option of syntax: `'setLogger()' => Instance::of('logger')`

I think it's more intuitive.

**Comment by samdark** [comment]
Yes, last proposal is good for me.

**Comment by SamMousa** [comment]
I still don't like it though; it is way more brittle than a closure.. 
Merging configs isn't something that should be done in my opinion, it can break stuff very easily. As a matter of fact i ran into that today.

**Comment by SamMousa** [comment]
#15089 merging configs is not guaranteed to work without making assumptions that you IMO shouldn't make.

**Comment by vtvz** [comment]
Another option of syntax more Yii-like: `'do setLogger' => Instance::of('logger')` as `as` for behaviors and `on` for events.

**Comment by rob006** [comment]
Calling setters from non-yii libraries could be simplified by https://github.com/yiisoft/yii2/issues/9748

**Comment by SamMousa** [comment]
@vtvz I don't agree with the basic assumption that configurations can or should be merged.
Let me give an example:
````
$config1 = [
'class' => 'A'
];
$config2 = [
'class' => 'subA',
'newProperty' => 15
];
$config3 = [
'class' => 'A'
];
````
Merging these will break since class `A` does not have property `$newProperty`. When merging configs you always need to know about the parent configuration. 
Also you get weird constructs like [these](https://github.com/yiisoft/yii2/blob/master/framework/helpers/UnsetArrayValue.php).

So in essence, when merging configs you need to know about the whole config tree, which kind of defeats the purpose of merging. (The only way to guarantee you know about the whole config is to redefine it and thus override everything, removing the need for a merge...)

**Comment by rob006** [comment]
> Let me give an example:

This is the same situation like extending class - you need to follow some rules to  make it work correctly. If you're overriding class in config, you should provide substitute (follow Liskov substitution principle) and consider changing this as BC break.

**Comment by SamMousa** [comment]
So then why use this "config" syntax at all? If we expect people who change configuration to know about the Liskov substitution principle would it not make more sense to just use actual code?
I'm not a fan of the array based configuration but I understand it is easier to use than actual configuration objects; however just merging them is risky and makes a lot of assumptions...

**Comment by rob006** [comment]
> So then why use this "config" syntax at all?

I'm not sure what is the alternative. If I *want to* merge configs, array syntax is the simplest and flexible enough solution - why change something that works?

**Comment by SamMousa** [comment]
That's a fair point, but the question remains, should we recommend this practice?

Anyway maybe that's more of a theoretical discussion, as long as we have a syntax with a space in the key it can never collide with anything so it won't cause people that don't use it any trouble.

**Comment by samdark** [comment]
Will be solved in 2.1 by using https://github.com/yiisoft/di. Won't be solved in 2.0.
